rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,453 @@
+This behavior is known in browsers as event bubbling.
+
+
+### Rewrite passes
+
+After the event pass, some flags may have been changed, and some values may have been invalidated and need to be recomputed.
+To address these invalidations, Masonry runs a set of **rewrite passes** over the tree:
+
+- **MUTATE** pass.
+- **UPDATE_TREE** pass.
+- **UPDATE_FOCUS** pass.
+- **UPDATE_DISABLED** pass.
+- **UPDATE_ANIM** pass.
+- **layout** pass.
+- **UPDATE_SCROLLS** pass.
+- **compose** pass.
+- **UPDATE_POINTER** pass.
+
+(The lowercase passes have methods with matching names in the Widget trait.)
+
+By default, each of these passes returns immediately, unless pass-dependent invalidation flags are set or work is requested.
+Each pass can generally request work for later passes; for instance, the MUTATE pass can invalidate the layout of a widget, in which case the `layout` pass will run on that widget and its children and parents.
+
+Passes may also request work for *previous* passes, in which case all rewrite passes are run again in sequence.
+For instance, the UPDATE_POINTER may change a widget's size, requiring another layout pass.
+
+To avoid infinite loops in those cases, the number of reruns has a static limit.
+If passes are still requested past that limit, they're delayed to a later frame.
+
+#### MUTATE pass
+
+The **MUTATE** pass runs a list of callbacks with mutable access to the widget tree.
+These callbacks can be queued with the `mutate_later()` method of various context types.
+
+"Mutable access" means that those callbacks are given a `WidgetMut` to the widget that requested them, something that is otherwise only accessible from the owner of the global `RenderRoot` object (see "External Mutation" section).
+
+#### UPDATE_XXX passes
+
+Update passes mostly run internal calculations.
+They compute if some widget's property has changed, and send it a matching `on_update_status` event (see "Status" section below).
+
+For instance, if a user presses `Tab` and the event isn't handled in a widget, the framework will run the `UPDATE_FOCUS` pass, which will automatically switch keyboard focus to the next focus-accepting widget.
+Both the previously-focused widget and the newly-focused widget will get an `on_update_status` call with relevant values.
+
+#### Layout pass
+
+The layout pass runs bidirectionally, passing constraints from the top down and getting back sizes and other layout info from the bottom up.
+
+It is subject to be reworked in the future to be closer to the semantics of web layout engines and the Taffy crate.
+
+Unlike with other passes, container widgets need their `Widget::layout()` method to call the `WidgetPod::layout()` method of their children.
+Not doing so is a logical bug and will panic when debug assertions are on.
+
+#### Compose pass
+
+The **compose** pass runs top-down and assigns transforms to children.
+Transform-only layout changes (e.g. scrolling) should request compose instead of requesting layout.
```

Comment:
Stashing in the sense it's currently used in the codebase still isn't implemented or documented.

Short answer is "compose could trigger a stash but it probably shouldn't".

> Would that require a mutate pass to be ran?

No, because stashing doesn't change the tree structure. A stashed widget still has the same parent, a parent with stash children still has them as children, etc.